src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -165,6 +177,15 @@ protected void connect(final SocketChannel channel) throws Exception {
 
     protected void startThreads(String threadName) throws InterruptedException {
         final CountDownLatch init = new CountDownLatch(2);
+
+        if (reader != null) {
+            reader.join(config.initTimeoutMillis / 2);
```

Comment:
What will happen if we'll leave this block by the timeout? To be honest, I don't got this change. The problem is that, say, a reader created before reconnection can wake up and set READING state in addition to RECONNECT? Whether your fix for StateHelper.acquire() (`if ((currentState & RECONNECT) > mask) { return false; }`) already fix it?

Reply:
> Whether your fix for StateHelper.acquire() (if ((currentState & RECONNECT) > mask) { return false; }) already fix it?

No, it doesn't. 
There's a rare race between old r/w threads and new ones. It's possible to reactivate RECONNECT state from the old threads and break down an initialization of the new threads.
This is fixed by two things:
1. the reconnect thread changes old thread for a new threads under RECONNECT state and then releases RECONNECT state.
2. R/W threads do an extra check in a reverse order: first they check the state is UNINIT and then check the current thread is the same active thread in the client. If so, threads are sure they are still the last generation threads. Otherwise, the threads skip an attempt to reconnect.

Reply:
The new way looks good.